In [116]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [117]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [118]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [119]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [120]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file

market_data_scaled = StandardScaler().fit_transform(df_market_data)


In [121]:
# Create a DataFrame with the scaled data
# Copy the crypto names from the original data

cols = ["price_change_percentage_24h",
        "price_change_percentage_7d",
        "price_change_percentage_14d",
        "price_change_percentage_30d",
        "price_change_percentage_60d",
        "price_change_percentage_200d",
        "price_change_percentage_1y"]

df_market_data_scaled = pd.DataFrame(StandardScaler().fit_transform(df_market_data), columns = cols)

# Set the coinid column as index

df_market_data_scaled["coin_id"] = df_market_data.index
df_market_data_scaled = df_market_data_scaled.set_index("coin_id")

#  Display sample data

df_market_data_scaled.sample(5)


price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                
monero                          0.262723                    1.792602   
binance-usd                     0.061339                   -0.706669   
tron                            0.130050                   -0.041018   
usd-coin                        0.034352                   -0.733026   
iota                            0.259097                    0.249508   

             price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                 
monero                          2.202665                     1.437842   
binance-usd                    -0.015321                    -0.058694   
tron                            0.147155                    -0.543776   
usd-coin                       -0.023140                    -0.065775   
iota                           -0.478953                    -0.218997   

             price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                  
monero                          0.893865                     -0.155893   
binance-usd                     0.004017                     -0.550003   
tron                            0.120116                     -0.241118   
usd-coin                        0.002925                     -0.550599   
iota                           -0.735815                     -0.329179   

             price_change_percentage_1y  
coin_id                                  
monero                        -0.167644  
binance-usd                   -0.281963  
tron                          -0.234014  
usd-coin                      -0.282232  
iota                          -0.285310

---

### Find the Best Value for k Using the Original Data.

In [122]:
# Create a list with the number of k-values from 1 to 11

k = list(range(1, 11))

In [123]:
# Create an empty list to store the inertia values

inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list

for i in k:
    k_model = KMeans(n_clusters = i, random_state = 2)
    k_model.fit(df_market_data_scaled)
    inertia.append(k_model.inertia_)

f:\Anaconda\lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
f:\Anaconda\lib\site-packages\sklearn\cluster\_kmeans.py:1382: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
f:\Anaconda\lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
f:\Anaconda\lib\site-packages\sklearn\cluster\_kmeans.py:1382: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
f:\A

In [149]:
# Create a dictionary with the data to plot the Elbow curve

elbow_data = {"k": k, "inertia": inertia}

# Create a DataFrame with the data to plot the Elbow curve

df_elbow = pd.DataFrame(elbow_data)

df_elbow

k     inertia
0   1  287.000000
1   2  195.820218
2   3  123.190482
3   4   79.022435
4   5   65.209390
5   6   53.617387
6   7   44.032965
7   8   37.614301
8   9   32.768762
9  10   28.967391

In [158]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.

ep1 = df_elbow.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve", 
    xticks=k
)

ep1

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** 4

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [126]:
# Initialize the K-Means model using the best value for k

model = KMeans(n_clusters = 4, random_state = 1)

In [127]:
# Fit the K-Means model using the scaled data

model.fit(df_market_data_scaled)

f:\Anaconda\lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
f:\Anaconda\lib\site-packages\sklearn\cluster\_kmeans.py:1382: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


KMeans(n_clusters=4, random_state=1)

In [128]:
# Predict the clusters to group the cryptocurrencies using the scaled data

clusters = model.predict(df_market_data_scaled)

# Print the resulting array of cluster values.

clusters

array([3, 3, 1, 1, 3, 3, 3, 3, 3, 1, 1, 1, 1, 3, 1, 3, 1, 1, 3, 1, 1, 3,
       1, 1, 1, 1, 1, 1, 3, 1, 1, 1, 0, 3, 1, 1, 2, 1, 1, 1, 1])

In [129]:
# Create a copy of the DataFrame

df_market_data_scaled_clustered = df_market_data_scaled

In [130]:
# Add a new column to the DataFrame with the predicted clusters

df_market_data_scaled_clustered["clusters"] = clusters

# Display sample data

df_market_data_scaled_clustered.sample(5)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                     
monero                               0.262723                    1.792602   
zcash                               -0.127467                    0.929119   
crypto-com-chain                     0.331280                   -1.614844   
litecoin                             0.077497                    0.334297   
ethlend                             -4.981042                   -0.045178   

                  price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                      
monero                               2.202665                     1.437842   
zcash                                0.677532                     0.223834   
crypto-com-chain                    -1.054521                    -0.729931   
litecoin                             0.858520                    -0.012646   
ethlend                             -1.206956                    -1.212126   

                  price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                       
monero                               0.893865                     -0.155893   
zcash                               -0.437068                     -0.265163   
crypto-com-chain                    -0.350155                     -0.022866   
litecoin                            -0.366477                     -0.486266   
ethlend                              0.047736                      4.632380   

                  price_change_percentage_1y  clusters  
coin_id                                                 
monero                             -0.167644         3  
zcash                              -0.214829         3  
crypto-com-chain                   -0.034570         1  
litecoin                           -0.292351         3  
ethlend                             6.088625         0

In [159]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.

plt1 = df_market_data_scaled_clustered.hvplot.scatter(
    x = "price_change_percentage_24h",
    y = "price_change_percentage_7d",
    by = "clusters",
    hover_cols = "coin_id"
)

plt1

:NdOverlay   [clusters]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [132]:
# Create a PCA model instance and set `n_components=3`.

pca = PCA(n_components = 3)

In [133]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.

df_market_data_scaled_fit = pd.DataFrame(pca.fit_transform(df_market_data_scaled))

# View the first five rows of the DataFrame. 

df_market_data_scaled_fit.head(5)

0         1         2
0 -1.585139  0.134394  0.545621
1 -1.322581  0.054494  1.115709
2  0.266335 -0.536409 -0.796946
3  0.257225 -0.592013 -0.646529
4 -2.952250  0.422372  1.267894

In [134]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.

print(pca.explained_variance_ratio_)
print(f"Total: {sum(pca.explained_variance_ratio_)}")

[0.37005408 0.32322221 0.19115222]
Total: 0.8844285111826466


#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** 0.8844285111826466

In [135]:
# Create a new DataFrame with the PCA data.
# Creating a DataFrame with the PCA data

df_market_data_scaled_fit

# Copy the crypto names from the original data

df_market_data_scaled_fit["coin_id"] = df_market_data_scaled_clustered.index

# Set the coinid column as index

df_market_data_scaled_fit = df_market_data_scaled_fit.set_index("coin_id")

# Display sample data

df_market_data_scaled_fit.sample(5)

0         1         2
coin_id                              
vechain  1.528493 -1.668876 -0.750235
dash     0.222043 -0.817068  0.122347
nem      0.408806  0.707042 -1.330139
tether   0.266335 -0.536409 -0.796946
ripple   0.257225 -0.592013 -0.646529

---

### Find the Best Value for k Using the PCA Data

In [136]:
# Create a list with the number of k-values from 1 to 11

k_2 = list(range(1, 11))

In [137]:
# Create an empty list to store the inertia values

inertia_2 = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list

for i in k_2:
    k_model = KMeans(n_clusters = i, random_state = 2)
    k_model.fit(df_market_data_scaled_fit)
    inertia_2.append(k_model.inertia_)


f:\Anaconda\lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
f:\Anaconda\lib\site-packages\sklearn\cluster\_kmeans.py:1382: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
f:\Anaconda\lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
f:\Anaconda\lib\site-packages\sklearn\cluster\_kmeans.py:1382: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(
f:\A

In [138]:
# Create a dictionary with the data to plot the Elbow curve

elbow_data_2 = {"k": k_2, "inertia": inertia_2}

# Create a DataFrame with the data to plot the Elbow curve

df_elbow_2 = pd.DataFrame(elbow_data_2)

df_elbow_2

k     inertia
0   1  287.007838
1   2  196.268358
2   3  111.972027
3   4   43.205879
4   5   32.330891
5   6   24.347640
6   7   18.717792
7   8   15.144404
8   9   12.252227
9  10   10.496204

In [160]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.

ep2 = df_elbow_2.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve", 
    xticks=k
)

ep2

:Curve   [k]   (inertia)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** 4


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** No

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [140]:
# Initialize the K-Means model using the best value for k

model_2 = KMeans(n_clusters = 4, random_state = 1)

In [141]:
# Fit the K-Means model using the PCA data

model_2.fit(df_market_data_scaled_fit)

f:\Anaconda\lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(
f:\Anaconda\lib\site-packages\sklearn\cluster\_kmeans.py:1382: UserWarning: KMeans is known to have a memory leak on Windows with MKL, when there are less chunks than available threads. You can avoid it by setting the environment variable OMP_NUM_THREADS=1.
  warnings.warn(


KMeans(n_clusters=4, random_state=1)

In [143]:
# Predict the clusters to group the cryptocurrencies using the PCA data

clusters_2 = model_2.predict(df_market_data_scaled_fit)

# Print the resulting array of cluster values.

clusters_2


array([1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 2, 1, 0, 0, 3, 0, 0, 0, 0])

In [144]:
# Create a copy of the DataFrame with the PCA data

df_market_data_scaled_fit_clustered = df_market_data_scaled_fit

# Add a new column to the DataFrame with the predicted clusters

df_market_data_scaled_fit_clustered["clusters"] = clusters_2

# Display sample data

df_market_data_scaled_fit_clustered.sample(5)

0         1         2  clusters
coin_id                                            
maker       -0.991935 -0.079763  0.741439         1
binance-usd  0.246128 -0.540957 -0.809982         0
litecoin    -1.347533 -0.132837  0.791063         1
zcash       -1.347135  0.106093  1.200966         1
bitcoin     -1.585139  0.134394  0.545621         1

In [163]:
# Create a scatter plot using hvPlot by setting 
# `x="PC1"` and `y="PC2"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.

plt2 = df_market_data_scaled_fit_clustered.hvplot.scatter(
    x = "0",
    y = "1",
    by = "clusters",
    hover_cols = "coin_id"
)

plt2

:NdOverlay   [clusters]
   :Scatter   [0]   (1,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [150]:
df_elbow["inertia pca"] = df_elbow_2["inertia"]
df_elbow

k     inertia  inertia pca
0   1  287.000000   287.007838
1   2  195.820218   196.268358
2   3  123.190482   111.972027
3   4   79.022435    43.205879
4   5   65.209390    32.330891
5   6   53.617387    24.347640
6   7   44.032965    18.717792
7   8   37.614301    15.144404
8   9   32.768762    12.252227
9  10   28.967391    10.496204

In [165]:
# Composite plot to contrast the Elbow curves

ep1 + ep2

:Layout
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k]   (inertia)

In [166]:
# Composite plot to contrast the clusters

plt1 + plt2

:Layout
   .NdOverlay.I  :NdOverlay   [clusters]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [clusters]
      :Scatter   [0]   (1,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** Clustering is stronger using the PCA data. Outliers are better classified into their own clusters
